### PR TITLE
Correct changing voices in part's linked staves

### PR DESCRIPTION
**Pull request by miiizen**
Resolves: https://github.com/musescore/MuseScore/issues/19180
This PR fixes the behaviour when changing the voice of a note on a stave with voices hidden through the 'Voices visible in the score' (N.b, should this not be 'Voices visible on the stave'?) options.
Screenshot 2024-07-22 at 09 55 14

When changing the voice of a note to one which is invisible on the current stave, the note is removed from the current stave and added to all those on which the voice is visible, as you might expect.  

When changing voice, we interpret the users choice as the _real underlying voice_, not the one currently displayed on the stave.  Eg. if the user has a stave configured to only have voice 2 visible, notes will appear as if in voice 1.  We do not allow changing the voice to voice 2 in this instance as it is already in voice 2.  I don't think this is ideal - there is definitely some design jank around this feature which would be great to clear up at some point.  However, this PR is focussed on fixing obviously broken behaviour.

### CODE REVIEW

**Review by oktophonie — APPROVED** [review]

**Comment by bkunda** [comment]
> N.b, should this not be 'Voices visible on the stave'?

I'd be inclined to go further: Voices visible on this stave

Good spot. "...in the score" is probably a little too vague, and would probably be more accurately put as "...from the score". 

But I think "...on this stave" relates even more directly to what the user sees in front of them at that moment.

File: src/engraving/dom/score.cpp [hunk 1]
```diff
@@ -5169,14 +5169,28 @@ void Score::changeSelectedNotesVoice(voice_idx_t voice)
...
             Chord* dstChord  = nullptr;
 
-            if (excerpt() && muse::key(excerpt()->tracksMapping(), dstTrack, muse::nidx) == muse::nidx) {
+            if (score->excerpt() && muse::key(score->excerpt()->tracksMapping(), dstTrack, muse::nidx) == muse::nidx) {
+                break;
+            }
+
+            if (score->excerpt()
+                && muse::key(score->excerpt()->tracksMapping(), chord->staffIdx() * VOICES + chord->voice(),
+                             muse::nidx) == chord->staffIdx() * VOICES + voice) {
```
Comment (mike-spa) [inline src/engraving/dom/score.cpp:5193]: I _kind_ of know what they mean but these conditions are quite hard to read/understand. Perhaps define a bool variable just before the if to hold the longer part of the statement and give it a name to make it clearer

File: src/engraving/dom/edit.cpp [hunk 2]
```diff
@@ -6471,19 +6474,43 @@ void Score::undoAddCR(ChordRest* cr, Measure* measure, const Fraction& tick)
...
+                    mappedTrack = (mappedTrack != muse::nidx) ? mappedTrack : track;
+                    if (staff->isVoiceVisible(strack % VOICES)) {
+                        mappedTracks.push_back(mappedTrack);
+                    }
+                }
+                for (track_idx_t track : mappedTracks) {
+                    // linkedPart : linked staves within same part/instrument.
+                    // linkedScore: linked staves over different scores via excerpts.
+                    const bool linkedPart  = linked && (staff != ostaff) && (staff->score() == ostaff->score());
+                    const bool linkedScore = linked && (staff != ostaff) && (staff->score() != ostaff->score());
```
Comment (mike-spa) [inline src/engraving/dom/edit.cpp:6491]: I know that this was there before, but while we're here, let's give these variables proper names such as isLinkedWithinSameScore and isLinkedToDifferentScore or something like that, rathert than use comments to clarify the meaning of poor names :)

File: src/engraving/dom/edit.cpp [hunk 3]
```diff
@@ -6471,19 +6474,43 @@ void Score::undoAddCR(ChordRest* cr, Measure* measure, const Fraction& tick)
...
+                        track_idx_t linkedTrack = staffIdx * VOICES + mappedTrack % VOICES;
+                        if (!staff->isVoiceVisible(strack % VOICES)) {
+                            continue;
+                        }
+                        tracks.push_back(linkedTrack);
                     } else if (!linkedPart && linkedScore) {
-                        if ((track >> 2) != staff->idx()) {
-                            track += (staff->idx() - (track >> 2)) * VOICES;
+                        // Staff in linked score
+                        if ((track >> 2) != staffIdx) {
```
Comment (mike-spa) [inline src/engraving/dom/edit.cpp:6504]: Ok, bit-shifting by 2 bits, in order to divide by 4, which is equal to VOICES, in order to get the staffIdx from a trackIdx, must be the most convoluted thing I've ever seen in this codebase 😅. Please let's take the opportunity to get rid of this and replace it with track / VOICES (or better use the track2staff function)

File: src/engraving/dom/edit.cpp (hunk 2 above)
Comment (miiizen) [reply src/engraving/dom/edit.cpp:6491]: I've done what I should have in the first place and made these changes in addition to refactoring this into a method within `Staff`, as this is duplicated in `undoAddElement`.

File: src/engraving/dom/staff.cpp [hunk 4]
```diff
@@ -124,6 +125,71 @@ Staff* Staff::findLinkedInScore(const Score* score) const
...
 
+std::vector<track_idx_t> Staff::getLinkedTracksInStaff(const Staff* linkedStaff, const track_idx_t originalTrack) const
+{
+    // For linked staves the length of staffList is always > 1 since the list contains the staff itself too!
+    const bool linked = staffList().size() > 1;
+
+    staff_idx_t staffIdx = linkedStaff->idx();
+
+    std::vector<track_idx_t> linkedTracks;
+    if (!linkedStaff->score()->excerpt()) {
```
Comment (mike-spa) [inline src/engraving/dom/staff.cpp:136]: I find `score()->isMaster()` more intuitive to read for these kind of master-vs-parts checks. We can do that in a next PR

**Review by mike-spa — APPROVED** [review]

**Pull request merged**